File: common/src/main/java/net/impleri/playerskills/integration/kubejs/PlayerSkillsPlugin.java
package net.impleri.playerskills.integration.kubejs;

import dev.architectury.event.events.common.LifecycleEvent;
import dev.latvian.mods.kubejs.KubeJSPlugin;
import dev.latvian.mods.kubejs.player.PlayerDataJS;
import dev.latvian.mods.kubejs.script.AttachDataEvent;
import dev.latvian.mods.kubejs.script.BindingsEvent;
import dev.latvian.mods.kubejs.script.ScriptType;
import net.impleri.playerskills.integration.kubejs.events.EventsBinding;
import net.impleri.playerskills.integration.kubejs.events.PlayerSkillChangedEventJS;
import net.impleri.playerskills.integration.kubejs.events.SkillsModificationEventJS;
import net.impleri.playerskills.integration.kubejs.events.SkillsRegistrationEventJS;
import net.impleri.playerskills.integration.kubejs.skills.BasicSkillJS;
import net.impleri.playerskills.integration.kubejs.skills.NumericSkillJS;
import net.impleri.playerskills.integration.kubejs.skills.SpecializedSkillJS;
import net.impleri.playerskills.integration.kubejs.skills.TieredSkillJS;
import net.impleri.playerskills.server.events.SkillChangedEvent;
import net.impleri.playerskills.variant.basic.BasicSkillType;
import net.impleri.playerskills.variant.numeric.NumericSkillType;
import net.impleri.playerskills.variant.specialized.SpecializedSkillType;
import net.impleri.playerskills.variant.tiered.TieredSkillType;
import net.minecraft.server.MinecraftServer;

public class PlayerSkillsPlugin extends KubeJSPlugin {

    private final PlayerSkillsKubeJSWrapper skillWrapper = new PlayerSkillsKubeJSWrapper();

    @Override
    public void init() {
        SkillChangedEvent.EVENT.register(this::onSkillChange);
        LifecycleEvent.SERVER_STARTING.register(this::onServerStart);
        Registries.SKILLS.addType(BasicSkillType.name.toString(), BasicSkillJS.Builder.class, BasicSkillJS.Builder::new);
        Registries.SKILLS.addType(NumericSkillType.name.toString(), NumericSkillJS.Builder.class, NumericSkillJS.Builder::new);
        Registries.SKILLS.addType(TieredSkillType.name.toString(), TieredSkillJS.Builder.class, TieredSkillJS.Builder::new);
        Registries.SKILLS.addType(SpecializedSkillType.name.toString(), SpecializedSkillJS.Builder.class, SpecializedSkillJS.Builder::new);
    }

    private void onServerStart(MinecraftServer minecraftServer) {
        // Trigger skills modification event
        new SkillsModificationEventJS(Registries.SKILLS.types).post(ScriptType.SERVER, EventsBinding.MODIFICATION);
    }

    @Override
    public void initStartup() {
        registerSkills();
    }

    @Override
    public void addBindings(BindingsEvent event) {
        event.add("PlayerSkills", skillWrapper);
    }

    @Override
    public void attachPlayerData(AttachDataEvent<PlayerDataJS> event) {
        event.add("skills", new MutablePlayerDataJS(event.parent().getMinecraftPlayer()));
    }

    private void registerSkills() {
        new SkillsRegistrationEventJS(Registries.SKILLS.types).post(ScriptType.STARTUP, EventsBinding.REGISTRATION);
    }

    private <T> void onSkillChange(SkillChangedEvent<T> event) {
        new PlayerSkillChangedEventJS<T>(event).post(ScriptType.SERVER, EventsBinding.SKILL_CHANGED, event.getSkill().getName().toString());
    }
}

File: common/src/main/kotlin/net/impleri/playerskills/integrations/kubejs/PlayerSkillsPlugin.kt
package net.impleri.playerskills.integrations.kubejs

import dev.architectury.event.events.common.LifecycleEvent
import dev.latvian.mods.kubejs.KubeJSPlugin
import dev.latvian.mods.kubejs.script.BindingsEvent
import dev.latvian.mods.kubejs.util.AttachedData
import net.impleri.playerskills.events.SkillChangedEvent
import net.impleri.playerskills.integrations.kubejs.mobs.MobEventsBinding
import net.impleri.playerskills.integrations.kubejs.mobs.MobRestrictionsEventJS
import net.impleri.playerskills.integrations.kubejs.skills.BasicSkillJS
import net.impleri.playerskills.integrations.kubejs.skills.MutablePlayerDataJS
import net.impleri.playerskills.integrations.kubejs.skills.NumericSkillJS
import net.impleri.playerskills.integrations.kubejs.skills.PlayerSkillChangedEventJS
import net.impleri.playerskills.integrations.kubejs.skills.PlayerSkillsKubeJSWrapper
import net.impleri.playerskills.integrations.kubejs.skills.SkillsEventsBinding
import net.impleri.playerskills.integrations.kubejs.skills.SkillsModificationEventJS
import net.impleri.playerskills.integrations.kubejs.skills.SkillsRegistrationEventJS
import net.impleri.playerskills.integrations.kubejs.skills.SpecializedSkillJS
import net.impleri.playerskills.integrations.kubejs.skills.TieredSkillJS
import net.impleri.playerskills.skills.basic.BasicSkillType
import net.impleri.playerskills.skills.numeric.NumericSkillType
import net.impleri.playerskills.skills.specialized.SpecializedSkillType
import net.impleri.playerskills.skills.tiered.TieredSkillType
import net.minecraft.server.MinecraftServer
import net.minecraft.world.entity.player.Player
import java.util.function.Consumer

class PlayerSkillsPlugin : KubeJSPlugin() {
  private val skillWrapper = PlayerSkillsKubeJSWrapper()
  override fun init() {
    LifecycleEvent.SERVER_STARTING.register(LifecycleEvent.ServerState { onServerStart(it) })

    SkillChangedEvent.EVENT.register(Consumer { onSkillChange(it) })

    Registries.SKILLS.addType(
      BasicSkillType.NAME.toString(),
      BasicSkillJS.Builder::class.java,
    ) { BasicSkillJS.Builder(it) }

    Registries.SKILLS.addType(
      NumericSkillType.NAME.toString(),
      NumericSkillJS.Builder::class.java,
    ) { NumericSkillJS.Builder(it) }

    Registries.SKILLS.addType(
      TieredSkillType.NAME.toString(),
      TieredSkillJS.Builder::class.java,
    ) { TieredSkillJS.Builder(it) }

    Registries.SKILLS.addType(
      SpecializedSkillType.NAME.toString(),
      SpecializedSkillJS.Builder::class.java,
    ) { SpecializedSkillJS.Builder(it) }
  }

  private fun onServerStart(server: MinecraftServer) {
    // Trigger skills modification event
    SkillsEventsBinding.MODIFICATION.post(SkillsModificationEventJS(Registries.SKILLS.types))
    MobEventsBinding.RESTRICTIONS.post(MobRestrictionsEventJS(server))
  }

  override fun registerEvents() {
    SkillsEventsBinding.GROUP.register()
    MobEventsBinding.GROUP.register()
  }

  override fun initStartup() {
    registerSkills()
  }

  override fun registerBindings(event: BindingsEvent) {
    event.add("PlayerSkills", skillWrapper)
  }

  override fun attachPlayerData(event: AttachedData<Player>) {
    event.add("skills", MutablePlayerDataJS(event.parent))
  }

  private fun registerSkills() {
    SkillsEventsBinding.REGISTRATION.post(SkillsRegistrationEventJS(Registries.SKILLS.types))
  }

  private fun <T> onSkillChange(event: SkillChangedEvent<T>) {
    SkillsEventsBinding.SKILL_CHANGED.post(event.skill.name, PlayerSkillChangedEventJS(event))
  }
}
